pkg/daemon/podprobe/worker.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podprobe
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/util"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/klog/v2"
+)
+
+// worker handles the periodic probing of its assigned container. Each worker has a go-routine
+// associated with it which runs the probe loop until the container permanently terminates, or the
+// stop channel is closed. The worker uses the probe Manager's statusManager to get up-to-date
+// container IDs.
+type worker struct {
+	// Channel for stopping the probe.
+	stopCh chan struct{}
+
+	// pod uid, container name, probe name
+	key probeKey
+
+	// Describes the probe configuration
+	spec *appsv1alpha1.ContainerProbeSpec
+
+	// The probe value during the initial delay.
+	initialValue appsv1alpha1.ProbeState
+
+	probeController *Controller
+
+	// The last known container ID for this worker.
+	containerID string
+	// The last probe result for this worker.
+	lastResult appsv1alpha1.ProbeState
+	// How many times in a row the probe has returned the same result.
+	resultRun int
+}
+
+// Creates and starts a new probe worker.
+func newWorker(c *Controller, key probeKey, probe *appsv1alpha1.ContainerProbeSpec) *worker {
+
+	w := &worker{
+		stopCh:          make(chan struct{}, 1), // Buffer so stop() can be non-blocking.
+		key:             key,
+		spec:            probe,
+		probeController: c,
+		initialValue:    appsv1alpha1.ProbeUnknown,
+	}
+
+	return w
+}
+
+// run periodically probes the container.
+func (w *worker) run() {
+	periodSecond := w.spec.PeriodSeconds
+	if periodSecond < 1 {
+		periodSecond = 1
+	}
+	probeTickerPeriod := time.Duration(periodSecond) * time.Second
+	probeTicker := time.NewTicker(probeTickerPeriod)
```

Comment:
If kubelet restarted, the probes could be started in rapid succession. Let the worker wait for a random portion of tickerPeriod before probing. plz refer to. pkg/kubelet/prober/prober.go for the example